File: code/common/Configuration.cs
using LanguageExt;
using Microsoft.Extensions.Configuration;

namespace common;

public static class ConfigurationModule
{
    public static string GetNonEmptyValue(this IConfiguration configuration, string key)
    {
        return configuration.GetOptionalNonEmptyValue(key)
                            .IfNoneThrow($"Configuration key '{key}' is missing or has a null, empty or whitespace value.");
    }

    public static Option<string> GetOptionalNonEmptyValue(this IConfiguration configuration, string key)
    {
        return configuration.GetOptionalSection(key)
                            .Map(section => section.Value)
                            .Where(value => string.IsNullOrWhiteSpace(value) is false);
    }

    public static string GetValue(this IConfiguration configuration, string key)
    {
        return configuration.GetOptionalValue(key)
                            .IfNoneThrow($"Could not find key '{key}' in configuration.");
    }

    public static Option<string> GetOptionalValue(this IConfiguration configuration, string key)
    {
        var section = configuration.GetSection(key);

        return section.Exists() ? section.Value : Prelude.None;
    }

    public static IConfigurationSection GetSection(IConfiguration configuration, string key)
    {
        return configuration.GetOptionalSection(key)
                            .IfNoneThrow($"Could not find section with key '{key}' in configuration.");
    }

    public static Option<IConfigurationSection> GetOptionalSection(this IConfiguration configuration, string key)
    {
        var section = configuration.GetSection(key);

        return section.Exists() ? Prelude.Some(section) : Prelude.None;
    }
}

File: code/common/Configuration.fs
namespace common

open Microsoft.Extensions.Configuration

[<RequireQualifiedAccess>]
module Configuration =
    let getOptionalSection key (configuration: IConfiguration) =
        let section = configuration.GetSection(key)

        if section.Exists() then
            Some section
        else
            None

    let getSection key configuration =
        configuration
        |> getOptionalSection key
        |> Option.defaultWith (fun () -> invalidOp $"Could not find section with key '{key}' in configuration.")

    let getOptionalValue key configuration =
        configuration
        |> getOptionalSection key
        |> Option.map (fun section -> section.Value)

    let getValue key configuration =
        configuration
        |> getSection key
        |> fun section -> section.Value
